plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/FormverFirBlock.kt

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright 2010-2025 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.formver.conversion
+
+import org.jetbrains.kotlin.fir.expressions.*
+import org.jetbrains.kotlin.fir.symbols.impl.FirFunctionSymbol
+import org.jetbrains.kotlin.formver.isFormverFunctionNamed
+
+private const val SPECIFICATION_CONDITION_ERROR =
+    "Only a single `preconditions` scope and a single `postconditions` scope is allowed as a function specification."
+
+private const val SPECIFICATION_LOOP_ERROR =
+    "Only a single `loopInvariants` scope is allowed in the loop."
+
+sealed interface FormverFirBlock {
+    val statements: FirBlock
+}
+
+data class LoopInvariantsBlock(override val statements: FirBlock) : FormverFirBlock
+data class PreconditionsBlock(override val statements: FirBlock) : FormverFirBlock
+data class PostconditionsBlock(override val statements: FirBlock) : FormverFirBlock
+
+interface FormverFirBlockBuilder {
+    fun complete(statements: FirBlock): FormverFirBlock
+}
+
+fun formverBlockBuilder(block: (FirBlock) -> FormverFirBlock) = object : FormverFirBlockBuilder {
+    override fun complete(statements: FirBlock) = block(statements)
+}
+
+
+fun FirFunctionSymbol<*>.blockBuilder(): FormverFirBlockBuilder? =
+    when {
+        // To avoid recursion, we rely on FIR to identify these blocks as early as possible.
```

Comment:
I think we should know when we can expect such a block, e.g. when looking into a function or loop body.  Why do we want to recurse over everything up front?